added new prop to Scoped component "postcss"

geoctrl opened the pull request:
Scoped component:
- new `postcss` prop to use instead of `css`
- uses classes instead of attributes
- each component tracks an `id` instead of the entire css string
- the counter is added as a property on the `<style>` tags

geoctrl commented on src/scoped.component.js:
```diff
@@ -9,21 +9,35 @@ const reactSupportsReturningArrays = !!ReactDOM.createPortal;
 export class Scoped extends React.Component {
   static propTypes = {
-    css: PropTypes.string.isRequired,
+    css: PropTypes.string,
```
the `css` prop is not required anymore because you can use the `postcss` prop instead

i'm checking for props in the constructor instead

geoctrl commented on src/scoped.component.js:
```diff
@@ -9,21 +9,35 @@ const reactSupportsReturningArrays = !!ReactDOM.createPortal;
+        if (this.props.css) {
+          elProps = { [this.state.kremlingAttrName]: this.state.kremlingAttrValue };
+        } else {
+          elProps = { className: `${this.props.postcss.id} ${child.props && child.props.className ? child.props.className : ''}` };
```
since we're using classes now instead of data attributes, we need to add our new scoped class along with the child's className prop

joeldenning replied:
Could you help me understand this? I at first was thinking that the postcss id was just for preventing the duplication of the same style element. But now it looks like it's used as part of our implementation of scoping? I'm not sure if that will work.

joeldenning commented on src/scoped.component.js:
```diff
@@ -127,6 +151,48 @@ export class Scoped extends React.Component {
   }
+
+  getPostcssSelector = (id) => {
+    return `style_${id.slice(1)}`;
```
why the `slice(1)`?

joeldenning commented on src/scoped.component.js:
```diff
@@ -127,6 +151,48 @@ export class Scoped extends React.Component {
+        this.getPostcssSelector(oldPostcss.id),
+        this.getPostcssSelector(newPostcss.id)
+      );
+      this.state.styleRef.innerHTML = newPostcss.styles;  
```
i think it might be possible here for some weird stuff to happen. If two components pass in the same post css id and styles but then one of them changes the post css id, shouldn't we look for a new `styleRef` instead of reusing the one we have. So that the other component doesn't accidentally get the new styles?

I might be getting confused here, but wanted to ask.

joeldenning commented on src/scoped.component.js:
```diff
@@ -127,6 +151,48 @@ export class Scoped extends React.Component {
+  }
+
+  reducePostcssCounter = () => {
+    this.state.styleRef.counter = this.state.styleRef.counter - 1;
```
nit, not important at all but just fyi: this is a perfect case for the decrement operator `this.state.styleRef.counter -= 1`.

joeldenning commented on src/scoped.component.js:
```diff
@@ -127,6 +151,48 @@ export class Scoped extends React.Component {
+  reducePostcssCounter = () => {
+    this.state.styleRef.counter = this.state.styleRef.counter - 1;
+    if (this.state.styleRef.counter === 0) {
+      this.state.styleRef.remove();
```
this doesn't work in IE11. See https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/remove. Instead, we should do `this.state.styleRef.parentNode.removeChild(this.state.styleRef)`

geoctrl commented:
- fixed the updating issues - if anything changes it'll remove the old `<style>` tag and create a new one
- changed the selectors back to attributes: `[namespace="id"]`
- added ability to pass in a namespace from the loader to kremling

joeldenning commented on src/scoped.component.js:
```diff
@@ -74,7 +71,7 @@ export class Scoped extends React.Component {
       this.doneWithCss();
     } else {
-      this.reducePostcssCounter();
+      this.doneWithPostss();
```
typo: `doneWithPostCss`

joeldenning commented on src/scoped.component.js:
```diff
@@ -152,47 +149,36 @@ export class Scoped extends React.Component {
-    // there's no style - create a new one and add it to the head
+    const kremlingAttrName = props.postcss.namespace || 'data-kremling';
+    const kremlingAttrValue = props.postcss.id;
+    let styleRef = document.head.querySelector(`[${kremlingAttrName}="${kremlingAttrValue}"]`);
```
if `this.state.styleRef` is defined, should we just use that?

joeldenning commented on src/scoped.component.js:
```diff
@@ -9,22 +9,30 @@ const reactSupportsReturningArrays = !!ReactDOM.createPortal;
       if (React.isValidElement(child)) {
         return React.cloneElement(child, {[this.state.kremlingAttrName]: this.state.kremlingAttrValue});
-      } else {
+        } else {
```
this indentation is weird
